attachment_synchronize/models/task.py

```diff
@@ -165,3 +170,8 @@ def run_import(self):
                         raise e
                     else:
                         new_env.cr.commit()
+
+    def _duplicated_file(self, filename):
+        if self.attachment_ids.search([('filename', '=', filename)]):
```

Comment:
Doing this, you read all attachments linked to the task, there is no need.
Replace by self.env['ir.attachment.metadata'].search()
Add task_id on the domain if you want to filter on attachment linked to this task (make sens to me)
Also, the field filename does not seem to exist. Maybe use the field datas_fname, we are sure the filename is inside : https://github.com/akretion/server-tools/blob/12-mig-external_file_location/attachment_synchronize/models/task.py#L88
You should probably add an index on the field in that case, for performance.

Also, as the system has to loop on all file names from the external system each time the cron runs, it may cause performance issue in the long run.
If your external app never delete or rename the files, you will end having issues I think

Reply:
well that's should not reading to all attachment, but searching on all attachment related to the task; probably filtered will be a better solution in this case, so kind of:
```
    if self.attachment_ids.filtered(lambda r: r.name == filename)
```
you're right, i meant to search for `name`. 

Yes, looping will be expensive in the long run, but actually it's better than process same file multiple times. 

I'll need to develop storage_backend_ftp, probably i could implement some kind of logic to list file modified later than last sync or similar.

Reply:
Well, I am not sure if filtered is a better option.
I mean, it depends if the external storage is sometimes cleaned or not...

Doing the filtered will be very expensive, because it needs to read all attachment linked to the task, which can be huge with time.
But once done, the result is put in cache.

Althoug a search will be really faster, in case you loop on a huge number of file, it will end up costing also...
So it will kind of depend of the use case.

About storage_backend_ftp, you should not need this as you could rename/move file with ftp right?

Anyway, I let you decide about the filtered or the search...

Reply:
Sorry i'm not getting the difference by searching using `attach_obj = self.env['ir.attachment.metadata']`in place of `attachment_ids`, could you please let me know. 

Probably would be faster if searching for a list of names in `self.env['ir.attachment.metadata']`and then get list difference and loop on that to import files.

> About storage_backend_ftp, you should not need this as you could rename/move file with ftp right?

What u mean by this?

Reply:
`attach_obj = self.env['ir.attachment.metadata']` is just an empty recordset
`while self.attachment_ids` is a recordset of the attachments linked to the task.
So both correspond to the same object, the difference is that the first one is empty and the second one is a recordset (which maybe huge).

So, if you don't actually need to read the linked attachment, it is better to go with an empty recordset.
But if you need to actually read the attachments (in case of filtered), then it is fine.

Anyway, if you do a `search`on a recordset, Odoo behave exatly the same as if you do a search on an empty recordset. Meaning that, if you do self.attachment_ids.search(...) Odoo will search among all attachments, and not only the one from the recordset (the one linked to the task). I am not sure I am very clear...

> Probably would be faster if searching for a list of names in self.env['ir.attachment.metadata']and then get list difference and loop on that to import files.

Probably a good compromise indeed.

`What u mean by this?`
What I said earlier. The module allow you to move/rename a file after importing.
You should be able to do this by ftp (if you have the write rights on the ftp server) in this case, the check duplicated feature is not needed !

Reply:
As far as i know, but at this point not anymore sure i got in the right way, searching by empty record set will search on all the table kind of:
```
     select * from ir_attachment_metadata
```
while `attachment_ids` should be already in memory object, so actually the only related to the task. I thought searching within this recordset will ended searching in memory instead of making query to the db, and this should be faster and over a smaller recordset. 

But sound like i misunderstood something 😬, that said no prob to use empty recordset if this is faster.

Reply:
` I thought searching within this recordset will ended searching in memory instead of making query to the db, and this should be faster and over a smaller recordset.`

No, a search will always do queries to the database.